tests/python_package_test/test_engine.py

```diff
@@ -1434,25 +1431,220 @@ def decreasing_metric(preds, train_data):
 
         def constant_metric(preds, train_data):
             return ('constant_metric', 0.0, False)
-
         # test that all metrics are checked (default behaviour)
+        params = {
+            'objective': 'regression',
+            'metric': 'None',
+            'verbose': -1
+        }
         gbm = lgb.train(params, lgb_train, num_boost_round=20, valid_sets=[lgb_eval],
                         feval=lambda preds, train_data: [decreasing_metric(preds, train_data),
                                                          constant_metric(preds, train_data)],
                         early_stopping_rounds=5, verbose_eval=False)
         self.assertEqual(gbm.best_iteration, 1)
-
         # test that only the first metric is checked
+        try:
+            gbm = lgb.train(dict(params, first_metric_only=True), lgb_train,
+                            num_boost_round=20, valid_sets=[lgb_eval],
+                            feval=lambda preds, train_data: [decreasing_metric(preds, train_data),
+                                                             constant_metric(preds, train_data)],
+                            early_stopping_rounds=5, verbose_eval=False)
+            self.assertTrue(False, "LightGBMError should be raised.")
+        except lgb.basic.LightGBMError as e:
+            self.assertEqual(e.args[0], '`first_metric_only` and `feval` are not available at the same time.')
+        # test that various combination of metrics.
+        params = {
+            'boosting_type': 'gbdt',
+            'objective': 'regression',
+            'learning_rate': 0.5,
+            'num_leaves': 5,
+            'metric': 'l2',
+            'verbose': -1,
+            'seed': 123,
+            'bagging_seed': 234,
+            'feature_fraction_seed': 456
+        }
         gbm = lgb.train(dict(params, first_metric_only=True), lgb_train,
-                        num_boost_round=20, valid_sets=[lgb_eval],
+                        num_boost_round=300, valid_sets=[lgb_eval],
+                        early_stopping_rounds=5, verbose_eval=False)
+        self.assertEqual(gbm.best_iteration, 34)
+
+        def metrics_combination_train_regression(metric_list, assumed_iteration, first_metric_only, assertion=True):
+            params = {
+                'boosting_type': 'gbdt',
+                'objective': 'regression',
+                'learning_rate': 0.05,
+                'num_leaves': 5,
+                'metric': metric_list,
+                'verbose': -1,
+                'seed': 123,
+                'bagging_seed': 234,
+                'feature_fraction_seed': 456
+            }
+            gbm = lgb.train(dict(params, first_metric_only=first_metric_only), lgb_train,
+                            num_boost_round=300, valid_sets=[lgb_eval],
+                            early_stopping_rounds=5, verbose_eval=False)
+            if assertion:
```

Comment:
Please describe why is this needed. Am I right, that it's used for determining `best_iter` in the first function call? Why don't hardcode `best_iter`?

Reply:
The reason why I didn't hardcoode the iteration number like the other test that the number was different depend on the environment. In my environment, if sklearn version is different, the the number is different. And also CI environment, depend on the test the number is different. But in case of CI environment, I don't understand the reason of difference...

### My local envirionment. (Mac OS)
scikit-learn-0.20.1 : 
-    l.1609 metrics_combination_cv('binary_logloss', 139, True, False) -> Failed. (AssertionError: 98 != 139)

scikit-learn-0.21.2 : 
-    l.1609 metrics_combination_cv('binary_logloss', 139, True, False) -> OK.


### CI environment
TASK=sdist PYTHON_VERSION=2.7
https://travis-ci.org/microsoft/LightGBM/jobs/540077858
-    l.1609 metrics_combination_cv('binary_logloss', 139, True, False) -> Failed. (AssertionError: 98 != 139)

TASK=regular PYTHON_VERSION=3.6
https://travis-ci.org/microsoft/LightGBM/jobs/540077857
-    l.1609 metrics_combination_cv('binary_logloss', 139, True, False) -> OK.

Reply:
@matsuken92 Oh, it's really weird! I'll fork your branch and try to figure out the reason.